Sources/PackageRegistry/RegistryManager.swift

```diff
@@ -0,0 +1,248 @@
+/*
+ This source file is part of the Swift.org open source project
+
+ Copyright (c) 2014 - 2020 Apple Inc. and the Swift project authors
+ Licensed under Apache License v2.0 with Runtime Library Exception
+
+ See http://swift.org/LICENSE.txt for license information
+ See http://swift.org/CONTRIBUTORS.txt for Swift project authors
+ */
+
+import PackageLoading
+import PackageModel
+
+import TSCBasic
+import TSCUtility
+
+import Foundation
+#if canImport(FoundationNetworking)
+import FoundationNetworking
+#endif
+
+public enum RegistryError: Error {
+    case invalidOperation
+    case invalidResponse
+    case invalidURL
+    case invalidChecksum(expected: String, actual: String)
+    case invalidPackage
+}
+
+public final class RegistryManager {
+    public static var configuration: URLSessionConfiguration = {
+        var configuration = URLSessionConfiguration.default
+        configuration.httpAdditionalHeaders = [
+            "Accept": "application/vnd.swift.registry.v1",
+            "User-Agent": "Swift Package Manager",
+        ]
+
+        return configuration
+    }()
+
+    private lazy var session = URLSession(configuration: Self.configuration)
+
+    let registryBaseURL: Foundation.URL
+
+    init(registryBaseURL: Foundation.URL) {
+        self.registryBaseURL = registryBaseURL
+    }
+
+    public class func discover(
+        for package: PackageReference,
+        timeoutInterval: TimeInterval = 1.0,
+        completion: @escaping (Result<RegistryManager, Error>) -> Void
+    ) {
+        guard case .remote(_, let urlString) = package else {
+            return completion(.failure(RegistryError.invalidPackage))
+        }
+
+        guard let url = URL(string: "https://\(urlString)") else {
+            return completion(.failure(RegistryError.invalidURL))
+        }
+
+        var request = URLRequest(url: url, cachePolicy: .returnCacheDataElseLoad, timeoutInterval: timeoutInterval)
+        request.httpMethod = "HEAD"
+
+        let session = URLSession(configuration: configuration)
+
+        let task = session.dataTask(with: request) { _, response, error in
+            if let error = error {
+                completion(.failure(error))
+            } else if let response = response as? HTTPURLResponse,
```

Comment:
Does the API spec provide guidance how registry should response to this kind of request?

Reply:
Yes, this is described under [4.1.1. Content negotiation](https://github.com/apple/swift-package-manager/blob/872ce5c7dfcf4df7cb8b636db4aa9707ba6e416e/Documentation/Registry.md#411-content-negotiation).